### PR TITLE
Interactive logs via `dagger run`

### PR DESCRIPTION
**Pull request by aluzzardi**
TODO:

- [x] Integrate with `dagger` CLI (e.g. `dagger run`)
  - It's currently a separate `main.go` that ingests a journal log (easier for development -- this task should be done last)
- [x] Pagination for the tree list
  - Currently all breaks loose when there's more items in the list that can be displayed on screen
- [x] Scrolling through logs is broken
- [x] "Follow" mode sometimes break (e.g. stuck on an item)
- [x] Color palette is inconsistent -- this should have a set of const lipgloss styles
- [x] General buggery -- screen flashes and so on

Screenshot 2023-02-02 at 5 41 59 PM

### CODE REVIEW

**Comment by vito** [comment]
fyi: doing a bit of necromancy on this branch; I'd like to use this for my services demo, since it makes it much easier to see everything that's going on

**Comment by sheldonhull** [comment]
I tried to give this a go, but mind giving an example of where it should work? I have a mage task running the dagger commands. 

I cloned/built from your repo:

```shell
mkdir -p ~/git/aluzzardi/dagger
git clone https://github.com/aluzzardi/dagger ~/git/aluzzardi/dagger
popd ~/git/aluzzardi/dagger
git checkout -b ui origin/ui
go build -o ~/go/bin/dagger-ui ./cmd/dagger-ui
```

Then tried a couple variations to invoke, including: 

- `dagger-ui go run mage.go -v build:appname`
- `dagger-ui mage -v build:appname`

I didn't dig through all of the PR, so if it's referenced let me know. Otherwise I'll wait. Figured I'd give this a try though. Cheers.

**Comment by vito** [comment]
@sheldonhull Thanks for trying it out! I recently changed it to send events over TCP, meaning your Dagger pipelines need to run with a dev `dagger session`, otherwise you'll just see a blank screen.

As a workaround you can do something like this:

```sh
env _EXPERIMENTAL_DAGGER_JOURNAL=/tmp/journal-$(date +%s) dagger-ui <command>
```

...which goes back to the tailing approach, which doesn't need the new session binary but might drop a few events due to partial reads/writes.

Or you can `go install ./cmd/dagger` from this same branch and set this before running `dagger-ui`:

```sh
export _EXPERIMENTAL_DAGGER_CLI_BIN=$HOME/go/bin/dagger # or wherever you installed it
```

...which will tell Dagger to use a custom binary instead of the one extracted from the engine image.

Let me know if this doesn't work for some reason.

/cc @d3rp3tt3

**Comment by aluzzardi** [comment]
@vito Damn, looking good!

**Comment by vito** [comment]
Latest round of polish since last update:

* The command being run is now showed as the "root" node, with its output shown in the "details" pane
* The Dagger engine is now started synchronously, rather than showing an empty UI that looks "stuck", we can more clearly see engine start lag time instead
* Got rid of the :hourglass: emoji (sorry) - it stuck out like a sore thumb once I tried on Mac
* Tree is now sorted in DAG/chronological order, no more strange ordering of items
* Using `--` to disambiguate flags is no longer required (`dagger run ./bass/test -i src=./` no longer fails with unknown `-i` flag)
* Support Home, End, Page Up, Page Down keys
* Pressing "up" or "down" (or more importantly Page Up/Page Down) while at/going past the top/bottom no longer wraps around to the other side
  * I think this is no longer necessary now that the other keybinds are there.
  * I like being able to just hold up/down if I'm being lazy.
* Handle super long titles better in more places, also fix a related panic

Still more refinements to do, but I'm inclined to prioritize merging this soon since some folks are already relying on it. It's not painful to rebase at least, but it's probably a good time to stop pixel-pushing and get it in more peoples' hands. wdyt @shykes @aluzzardi?

File: cmd/dagger/exec_unix.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	// cmd.SysProcAttr.Pdeathsig = syscall.SIGKILL
+	cmd.SysProcAttr.Setpgid = true
+
+	cmd.Cancel = func() error {
+		pgid, err := syscall.Getpgid(cmd.Process.Pid)
+		if err != nil {
+			return err
+		}
+		return syscall.Kill(-pgid, syscall.SIGKILL)
```
Comment (sipsma) [inline cmd/dagger/exec_unix.go:24]: iiuc this is what we'll send if `dagger run` gets `SIGINT`, right? If so, sending `SIGKILL` right away seems a little aggressive

File: cmd/dagger/main.go [hunk 2]
```diff
@@ -25,6 +29,7 @@ func init() {
 
 	rootCmd.PersistentFlags().StringVar(&workdir, "workdir", ".", "The host workdir loaded into dagger")
 	rootCmd.PersistentFlags().BoolVar(&debugLogs, "debug", false, "show buildkit debug logs")
+	rootCmd.PersistentFlags().StringVar(&cpuprofile, "cpuprofile", "", "collect CPU profile to path, and trace at path.trace")
```
Comment (sipsma) [inline cmd/dagger/main.go:32]: 🆒

**Review by sipsma — APPROVED** [review]
Only looked at the non-tui stuff in depth, all the screenshots of the TUI look great though, makes sense to get it out and keep iterating I think.

A few nits on the other stuff but nothing blocking

File: cmd/dagger/exec_unix.go (hunk 1 above)
Comment (vito) [reply cmd/dagger/exec_unix.go:24]: Also fires if you press `q`.

I can be talked out of it, but this is intentional. Telling a command to exit and then having it hang is incredibly frustrating, especially when it's a program that completely takes over your terminal, and especially when it happens constantly as part of a tight feedback loop. IMO everything should just try to be durable to a `kill -9` instead, since you often have to deal with nasty exit states anyway.

I figure we can get away with especially in this context where just about everything should be containerized and easy to re-run with minimal "pollution" from abrupt exits. But I'll watch for feedback here.

Comment (sipsma) [reply cmd/dagger/exec_unix.go:24]: I think the problem is that this is used for `dagger run`, so the command the user is executing may actually require not getting sigkilled if it needs to do some cleanup. I think it's probably at least a good idea to make this configurable at some point

Comment (vito) [reply cmd/dagger/exec_unix.go:24]: I'll can switch it to something softer and see if anything starts hanging for me, maybe I'm chasing ghosts.

I get that people like to do all sorts cleanup when interrupted, I just have an unusually extremist position that they shouldn't do that. :sweat_smile: It's awful UX (slowness, timeouts, stuck states, bricked terminal) and doesn't prevent you from having to deal with leaked resources anyway. (What if the dev was impatient that day and just spammed Ctrl+C? What if your laptop battery died? What if the application panicked and it wasn't trapped? What if your pipelines call `os.Exit(1)`? etc.)

In short I've never met a cleanup handler I liked. I'd rather just have a cleanup script.

Comment (vito) [reply cmd/dagger/exec_unix.go]: Changed it to SIGTERM + 10 second timeout (`--cleanup-timeout`) => SIGKILL but let me know if you've got a different preference for the default!

File: cmd/dagger/main.go (hunk 2 above)
Comment (gerhard) [reply cmd/dagger/main.go:32]: 🤯

**Comment by gerhard** [comment]
This just blew my mind:

image

image


Will continue reviewing, but this was worth stopping for.

**Review by gerhard — COMMENTED** [review]
I also think that we should put this behind and `_EXPERIMENTAL_X` feature flag, and mature it for a few weeks before we release it as a new minor. Yes, this is that good!

**Comment by vito** [comment]
> When we first merge this as an experimental feature (with corresponding flag / env variable)? This buys us a little extra time to iron out the kinks, and then it can be material for the next minor release (definitely big enough to justify it).

@shykes Happy to make this change if we want.

The current state is that this only affects `dagger run`. My understanding is `dagger run` isn't commonly used so it already felt sufficiently hidden, but I could have that wrong.

Looking through our docs, we reference it in a just two places, and both are somewhat advanced topics:

* https://github.com/dagger/dagger/blob/f58f964d3e0cf8a722dd9b5541bf63cecce156d8/core/docs/d7yxc-operator_manual.md?plain=1#L90
* https://github.com/dagger/dagger/blob/f58f964d3e0cf8a722dd9b5541bf63cecce156d8/docs/current/api/254103-build-custom-client.md?plain=1#L117

I can add an env var if we want to formally declare it experimental anyway.

**Comment by gerhard** [comment]
Follow-up feature request: horizontal split.

image

Too many tasks to fit in ~900px (full height is 1800px):

image

**Comment by gerhard** [comment]
cc @sheldonhull re `mage`:

image

**Comment by vito** [comment]
Added `_EXPERIMENTAL_DAGGER_TUI=1` env var and tested it out locally. If not set, `dagger run` falls back to essentially what it was doing before.

**Comment by gerhard** [comment]
I ran this with our `internal/mage`, and I'm pretty sure that the `-w` flag is not passed through correctly:

image

Am I doing something wrong?

**Comment by vito** [comment]
@gerhard The issue there is likely that the workdir state is fixed at `dagger run` time. The `mage` CLI changes the workdir with `-w` but by then the Dagger session already has an opinion of its workdir, which is what is tied to host paths.

**Comment by shykes** [comment]
> The current state is that this only affects `dagger run`. My understanding is `dagger run` isn't commonly used so it already felt sufficiently hidden, but I could have that wrong.

You're absolutely right. The experimental flag is not really needed from a UX perspective, since as you say, `dagger run` in its current undocumented form, acts as an experimental flag of its own :)

My only concern is with getting this feature the limelight it deserves, which means making it part of a minor release. From that point of view, I see two options:

1. No experimental flag, add it to the 0.5 release train ("dynamic secrets + interactive logs")
2. Experimental flag, release 0.6 in two weeks ("interactive logs!"). In this case the flag is strictly to make the 0.6 announcement not a lie (some extra maturing doesn't hurt either of course)

File: cmd/dagger/run.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ DAGGER_SESSION_PORT and DAGGER_SESSION_TOKEN will be convieniently injected auto
 }
 
 var waitDelay time.Duration
+var useShinyNewTUI bool = os.Getenv("_EXPERIMENTAL_DAGGER_TUI") == "1"
```
Comment (gerhard) [inline cmd/dagger/run.go:49]: Any value other than `""` should enable it IMO.

**Comment by gerhard** [comment]
@shykes 2 sounds great to me, especially since we have a few known edge cases, and a bunch of improvements that would be nice to look into over the next 2 weeks. Also, this is as close as we will get to an Easter egg. I really like the idea of `_EASTER_EGG=1` having the same effect as `_EXPERIMENTAL_DAGGER_TUI` on `dagger run`.

**Comment by gerhard** [comment]
Just confirmed that `_EXPERIMENTAL_DAGGER_TUI=1` behaves as expected:

image

File: cmd/dagger/main.go (hunk 2 above)
Comment (gerhard) [reply cmd/dagger/main.go:32]: Just curious, how do you work with Go `*.trace` files @vito ?

File: cmd/dagger/run.go (hunk 3 above)
Comment (vito) [reply cmd/dagger/run.go:49]: You're right - I blame Copilot. :)

File: internal/tui/editor.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	// also, editors that might not take kindly to being run from within a
+	// terminal in *another* editor
+	"vim",
+	"vi",
+	"emacs",
+	"helix",
+
+	// everyone has these, right?
+	"nano",
+	"pico",
```
Comment (gerhard) [inline internal/tui/editor.go:32]: 🤣

**Review by gerhard — APPROVED** [review]
💙 🚀 🪺

File: cmd/dagger/main.go (hunk 2 above)
Comment (vito) [reply cmd/dagger/main.go:32]: I was using this: https://github.com/felixge/traceutils

File: internal/tui/editor.go (hunk 4 above)
Comment (sheldonhull) [reply internal/tui/editor.go:32]: I didn't know Pico existed. 😆
I'm so used to vscode I feel like half a dev every time the magic of nvim being used is demoed. 😆

I did have an ergodox with blank keycaps to work on my geek cred but decided it was too much for my feeble brain.
